server/src/main/java/org/opensearch/index/query/functionscore/TermFrequencyFunction.java

```diff
@@ -0,0 +1,152 @@
+                case SUM_TOTAL_TERM_FREQ:
+                    return new SumTotalTermFreq(field, term, docId, context);
+                default:
+                    throw new IllegalArgumentException("Unsupported function: " + functionName);
+            }
+        }
+    }
+
+    /**
+     * TermFreqFunction computes the term frequency in a field.
+     */
+    public static class TermFreqFunction extends TermFrequencyFunction {
+
+        public TermFreqFunction(String field, String term, int docId, Map<Object, Object> context) {
+            super(field, term, docId, context);
+        }
+
+        @Override
+        public Integer execute(LeafReaderContext readerContext) throws IOException {
+            TermFreqValueSource valueSource = new TermFreqValueSource(field, term, field, BytesRefs.toBytesRef(term));
+            return valueSource.getValues(null, readerContext).intVal(docId);
+        }
+    }
+
+    /**
+     * TFFunction computes the term frequency-inverse document frequency (tf-idf) in a field.
+     */
+    public static class TFFunction extends TermFrequencyFunction {
+
+        public TFFunction(String field, String term, int docId, Map<Object, Object> context) {
+            super(field, term, docId, context);
+        }
+
+        @Override
+        public Float execute(LeafReaderContext readerContext) throws IOException {
+            TFValueSource valueSource = new TFValueSource(field, term, field, BytesRefs.toBytesRef(term));
+            return valueSource.getValues(context, readerContext).floatVal(docId);
+        }
+    }
+
+    /**
+     * TotalTermFreq computes the total term frequency in a field.
+     */
+    public static class TotalTermFreq extends TermFrequencyFunction {
+
+        public TotalTermFreq(String field, String term, int docId, Map<Object, Object> context) {
+            super(field, term, docId, context);
+        }
+
+        @Override
+        public Long execute(LeafReaderContext readerContext) throws IOException {
+            TotalTermFreqValueSource valueSource = new TotalTermFreqValueSource(field, term, field, BytesRefs.toBytesRef(term));
+            valueSource.createWeight(context, (IndexSearcher) context.get("searcher"));
+            return valueSource.getValues(context, readerContext).longVal(docId);
+        }
+    }
+
+    /**
+     * SumTotalTermFreq computes the sum of total term frequencies within a field.
+     */
+    public static class SumTotalTermFreq extends TermFrequencyFunction {
+
+        public SumTotalTermFreq(String field, String term, int docId, Map<Object, Object> context) {
+            super(field, term, docId, context);
+        }
+
+        @Override
+        public Long execute(LeafReaderContext readerContext) throws IOException {
+            SumTotalTermFreqValueSource valueSource = new SumTotalTermFreqValueSource(field);
+            valueSource.createWeight(context, (IndexSearcher) context.get("searcher"));
```

Comment:
This code (and the code for other functions) gets executed on every doc, right?

If so, we'll need to figure out a way of caching these per ScoreScript. I think we could move the docId into the `execute` method parameter and the `leafReadContext` into the constructor/factory. Then the function constructors would do most of the heavy lifting (creating value sources + weights as needed), and the `execute` method can load the value for the given doc ID.